File: p098m/valid_bst.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
const Tree = require('binary_tree');
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBSTV1 = function (root) {
  // A tree is BST if
  // 1. its left is a BST and root is greater than any of left
  // 2. its right is a BST and root is less than any of the right
  const recIsBSTHelper = function (root, minVal, maxVal) {
    if (root === null) return true;
    if (root.val <= minVal || root.val >= maxVal) return false;
    return (
      recIsBSTHelper(root.left, minVal, root.val) &&
      recIsBSTHelper(root.right, root.val, maxVal)
    );
  };
  return recIsBSTHelper(root, Number.MIN_SAFE_INTEGER, Number.MAX_SAFE_INTEGER);
};
// The above solution relies on INT_MIN and INT_MAX which apparently doesn't scale to long, double etc.
// This version traverses the tree in-order, checking if they form increasing order
var isValidBSTV2 = function (root) {
  let prev = null; // closure of the prev node, which must be less than the current node
  let recValidate = function (node) {
    if (node == null) return true;
    if (!recValidate(node.left)) return false;
    if (prev && prev.val >= node.val) return false;
    prev = node;
    return recValidate(node.right);
  };
  return recValidate(root);
};
// TEST
[
  ['0,#,-1,#,#', false],
  ['0,#,#', true],
  ['1,#,#', true],
  ['1,2,#,#,#', false],
  ['2,#,1,#,#', false],
  ['2,1,#,#,#', true],
  ['1,2,#,#,3,#,#', false],
  ['2,1,#,#,3,#,#', true],
  ['5,1,#,#,4,3,#,#,6,#,#', false],
].forEach(([tree, expected]) => {
  const actual = isValidBSTV1(Tree.deserialize(tree));
  console.log('Is', tree, 'a valid BST? ->', actual);
  console.assert(actual === expected);
  console.assert(isValidBSTV2(Tree.deserialize(tree)) === actual);
});

File: p098m/valid_bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from local_packages.binary_tree import TreeNode


class Solution:
    def isValidBST(
        self, root: TreeNode, lb: int = float("-Inf"), ub: int = float("Inf")
    ) -> bool:
        if not root:
            return True
        if root.val <= lb or root.val >= ub:
            return False
        return self.isValidBST(root.left, lb, root.val) and self.isValidBST(
            root.right, root.val, ub
        )


# TESTS
for tree, expected in [
    ("0,#,-1,#,#", False),
    ("0,#,#", True),
    ("1,#,#", True),
    ("1,2,#,#,#", False),
    ("2,#,1,#,#", False),
    ("2,1,#,#,#", True),
    ("1,2,#,#,3,#,#", False),
    ("2,1,#,#,3,#,#", True),
    ("5,1,#,#,4,3,#,#,6,#,#", False),
]:
    sol = Solution()
    actual = sol.isValidBST(TreeNode.deserialize(tree))
    print("Is", tree, "a valid BST? ->", actual)
    assert actual == expected
